src/async-wrap.cc

```diff
@@ -280,33 +280,23 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
                         Local<Value> parent, void* arg) {
   Local<Context> context = promise->CreationContext();
   Environment* env = Environment::GetCurrent(context);
-  if (type == PromiseHookType::kInit) {
-    // Unfortunately, promises don't have internal fields. Need a surrogate that
-    // async wrap can wrap.
-    Local<Object> obj =
-      env->async_hooks_promise_object()->NewInstance(context).ToLocalChecked();
-    PromiseWrap* wrap = new PromiseWrap(env, obj);
-    v8::PropertyAttribute hidden =
-      static_cast<v8::PropertyAttribute>(v8::ReadOnly
-                                         | v8::DontDelete
-                                         | v8::DontEnum);
-    promise->DefineOwnProperty(context,
-              env->promise_wrap(),
-              v8::External::New(env->isolate(), wrap),
-              hidden).FromJust();
-    // The async tag will be destroyed at the same time as the promise as the
-    // only reference to it is held by the promise. This allows the promise
-    // wrap instance to be notified when the promise is destroyed.
-    promise->DefineOwnProperty(context,
-              env->promise_async_tag(),
-              obj, hidden).FromJust();
+  PromiseWrap* wrap = Unwrap<PromiseWrap>(promise);
+  if (type == PromiseHookType::kInit ||
+      wrap == nullptr) {
+    uint32_t stored_init_field;
+    if (type != PromiseHookType::kInit) {
+      // Do not emit an init event.
+      stored_init_field = env->async_hooks()->fields()[AsyncHooks::kInit];
+      env->async_hooks()->fields()[AsyncHooks::kInit] = 0;
```

Comment:
well, this is a fun hack :) – I'm not sure I approve, would a default parameter in `AsyncWrap::AsyncWrap` and `AsyncWrap::AsyncReset` be that difficult?

@Fishrock123 thoughts?